internal/venice-client-common/src/main/java/com/linkedin/venice/schema/SchemaUtils.java

```diff
@@ -100,4 +105,91 @@ public static GenericRecord createGenericRecord(Schema originalSchema) {
+    if (replicatedSchema.getType().equals(RECORD)) {
+      for (Schema.Field field: replicatedSchema.getFields()) {
+        if (field.schema().getType().equals(MAP)) {
+          annotateMapSchema(field.schema());
+        }
+      }
+    }
+    return replicatedSchema;
+  }
+
+  /**
+   * Annotate all the top-level map field of the partial update schema to use Java String as key. This method will make
+   * sure field update and collection merging operations of a top-level map field are annotated.
+   * @param schema the input partial update schema to be annotated.
+   * @return Annotated partial update schema.
+   */
+  public static Schema annotateStringMapInPartialUpdateSchema(Schema schema) {
+    // Create duplicate schema here in order not to create any side effect during annotation.
+    Schema replicatedSchema = AvroSchemaParseUtils.parseSchemaFromJSONStrictValidation(schema.toString());
+    if (replicatedSchema.getType().equals(RECORD)) {
+      for (Schema.Field field: replicatedSchema.getFields()) {
+        if (field.schema().isUnion()) {
+          for (Schema unionBranchSchema: field.schema().getTypes()) {
+            // Full update request for Map field.
+            if (unionBranchSchema.getType().equals(Schema.Type.MAP)) {
+              annotateMapSchema(unionBranchSchema);
+            } else if (unionBranchSchema.getType().equals(RECORD)) {
+              for (Schema.Field updateOpField: unionBranchSchema.getFields()) {
+                Schema.Type updateOpFieldType = updateOpField.schema().getType();
+                if (updateOpFieldType.equals(MAP)) {
+                  annotateMapSchema(updateOpField.schema());
+                } else if (updateOpFieldType.equals(ARRAY)
+                    && updateOpField.schema().getElementType().getType().equals(STRING)) {
+                  annotateStringArraySchema(updateOpField.schema());
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    return replicatedSchema;
+  }
+
+  /**
+   * Create a new {@link SchemaEntry} for value schema annotation.
+   * @param schemaEntry Input {@link SchemaEntry}, containing the value schema to be annotated.
+   * @return Annotated value schema in a newly created {@link SchemaEntry}
+   */
+  public static SchemaEntry getAnnotatedStringMapValueSchemaEntry(SchemaEntry schemaEntry) {
+    if (schemaEntry == null) {
+      return null;
+    }
+    Schema annotatedSchema = annotateStringMapInValueSchema(schemaEntry.getSchema());
+    return new SchemaEntry(schemaEntry.getId(), annotatedSchema);
+  }
+
+  /**
+   * Create a new {@link DerivedSchemaEntry} for partial update schema annotation.
+   * @param schemaEntry Input {@link DerivedSchemaEntry}, containing the partial update schema to be annotated.
+   * @return Annotated partial update schema in a newly created {@link DerivedSchemaEntry}
+   */
+  public static DerivedSchemaEntry getAnnotatedStringMapDerivedSchemaEntry(DerivedSchemaEntry schemaEntry) {
```

Comment:
Can we update the function name here as similar as `annotateStringMapInPartialUpdateSchema`?

Reply:
I can do this but what I put here is "DerivedSchemaEntry" is the class name, which is incorrect, and I would like to change it together with a new PR.